Integration_test  package support and null-safety

jonsamwell opened the pull request:
Note that null safety cannot yet be achived in the flutter stable branch due to flutter_driver and integration_test not being null safe themselves see https://github.com/flutter/flutter/issues/71379 & https://github.com/flutter/flutter/issues/80293

jonsamwell committed:
feat(integration_test): added concept of an app driver adapter putting a facade over the Flutter driver and WidgetTester so driving the app becomes agnostic of implementation

jonsamwell committed:
Update steps to work with new app driver adapter

jonsamwell committed:
feat(integration_test): add a WidgetTester app driver adapter

jonsamwell committed:
feat(project): re-organised project structure

jonsamwell committed:
test(config): fixed tests

jonsamwell committed:
chore(project): move structure changes and remove FlutterDriverUtils as this functionality is now in the app driver adapters

jonsamwell committed:
feat(integration_test): first pass of code generation and gherkin tests running using the integration_test library

jonsamwell committed:
chore(libs): referencing the update dart_gherkin branch

jonsamwell committed:
feat(dependencies): added in calls to reporters and hooks

jonsamwell committed:
feat(integration_test): more progress with reporting

jonsamwell committed:
chore(lint): fixes

jonsamwell committed:
feat(reporters): got the json reporter working and results saving

jonsamwell committed:
chore(lib): updated changelog

jonsamwell committed:
fix(flutter_driver): various changes to get the flutter_driver part of the lib working so need to remove references to flutter_test / dart:ui

jonsamwell committed:
feat(integration_test): added much better example app that resets state before each test

jonsamwell committed:
fix(reports): fixed default report path

jonsamwell committed:
fix(generation): escape text in strings

jonsamwell committed:
fix(adapters): fixed issue with isPresent for test widget adapter

Commit:
fixes for changelog

jonsamwell committed:
Merge pull request #109 from vrnvorona/integration_test__package_support

fixes for changelog

jonsamwell committed:
fix(appDriver): fixed issue when trying to scroll using the WidgetTester adapter

jonsamwell committed:
fix(adapter): fix scrollUntilVisible in adatpers

Commit:
feature folder fix in changelog

jonsamwell committed:
Merge pull request #110 from vrnvorona/integration_test__package_support

feature folder fix in changelog

jonsamwell committed:
feat(drivers): added typed driver flutter world

jonsamwell committed:
feat(expressions): now evaluating tag expressions before running a scenario

jonsamwell committed:
fix(world): fixed drvier world typing

jonsamwell committed:
fix(worlds): update to typed world

jonsamwell committed:
fix(world): updated typed flutter world

jonsamwell committed:
fix(steps): wait for app to settle after page back

jonsamwell committed:
fix(steps): added pump invokes to some steps

jonsamwell committed:
fix(integration_test): ensure tearDownAll is always run

jonsamwell committed:
fix(widget): fixed screenshot using widget tester

jonsamwell committed:
fix(driver): ensure timeout on tap is respected

jonsamwell committed:
fix(steps): ensure wait for element step actually waits rather than testing and returning immediately

jonsamwell committed:
fix(steps): ensure app is pumped after step actions

jonsamwell committed:
fix(api): expose widget tester flutter world

jonsamwell committed:
feat(null-safety): first pass of null safety

jonsamwell committed:
feat(null-satefy): more lib version tweaks

jonsamwell commented:
This branch doesn't actually run yet due to various null-safety incompatibilities with flutter_driver and integration_test

tshedor reviewed (commented):
The `integration_test` stuff is exciting. We've been evaluating ways to make our test runs more performant, and one of those was using Gherkin -> Flutter test. So this is super cool stuff.

However, I think it's too much to couple with the null safety release. I think this is a release of its own as it's a massive feature set and a radically different implementation change. This would be a v3 to null safety's v2. 

As the maintainer of a package that relies extensively on `build_runner`, it's the tip of the iceberg of an extremely fickle ecosystem (`source_gen`, `source_gen_test`, `build_config`, `analyzer`). The Dart core team maintains these rapidly or doesn't maintain them at all. json_serializable always releases with the changes, but it's a challenge for community packages to keep up. There isn't another option to `build_runner`, so I'm not discouraging its use here, but I feel like I should provide some warning. And encourage null safety being a separate release instead of coupling it with file generation and `build_runner`.

jonsamwell commented:
@tshedor really appreciate  your thoughts here!!! I'll go over them in more detail tomorrow, but you are right this does feel like it should be released after null-safety

tshedor commented on CHANGELOG.md:
````diff
@@ -1,3 +1,121 @@
+}
+```
+4. Create a folder call `integration_test` this will eventually contain all your Gherkin feature files and the generated test files.
+5. Add the following file (and folder) `integration_test\features\counter.feature` with the following below contents.  This is a basic feature file that will be transform in to a test file that can run a test against the sample app.
````
I also think this should be broken up. This is part example / part required set up. There should be two distinct instruction sets: one as a tutorial to follow for new implementations and the other a migration guide for existing implementations.

tshedor commented on .flutter-plugins:
```diff
@@ -0,0 +1,2 @@
+# This is a generated file; do not edit or check into version control.
```
This file should be ignored

tshedor commented on .flutter-plugins-dependencies:
```diff
@@ -0,0 +1 @@
+{"info":"This is a generated file; do not edit or check into version control.","plugins":{"ios":[{"name":"integration_test","path":"C:\\\\Google\\\\flutter\\\\packages\\\\integration_test\\\\","dependencies":[]}],"android":[{"name":"integration_test","path":"C:\\\\Google\\\\flutter\\\\packages\\\\integration_test\\\\","dependencies":[]}],"macos":[],"linux":[],"windows":[],"web":[]},"dependencyGra ...
```
This should also be ignored

jonsamwell committed:
chore(null-safety): more fixes
fix(adapters): fixed an issue with the pumpAndSettle method

jonsamwell committed:
chore(dependencies): updates to pubspec to allow alpha release

jonsamwell committed:
chore(rename): renamed rawDriver -> nativeDriver

jonsamwell committed:
chore(lint): various code liniting changes

jonsamwell committed:
chore(gen): regenerated tests

jonsamwell committed:
chore(libs): updated to latest dart_gherkin

jonsamwell committed:
fix(#150 #141 #128): fixed handling of json in multiline string and improved and added example of generating html report

jonsamwell committed:
chore(lib): fixed lib constraint warning

jonsamwell committed:
chore(debug): removed debug code

jonsamwell committed:
fix(runner): Ensure scenario support files (world etc) as always disposed ensure when test throws error

jonsamwell committed:
- BREAKING CHANGE: Made `appMainFunction` return a `Future<void>` so it can be async
- Fix: #159: Swipe step not working
- Ensure Hook.onBeforeRun is called before the run starts
- Set Frame policy- defaults to `LiveTestWidgetsFlutterBindingFramePolicy.benchmarkLive` to slightly improve performance

jonsamwell committed:
- Fix: #165: Empty .feature files causing void functions which get compiled out at runtime and cause errors
- Fix: #162: Incorrect feature name in HTML reports - many thanks to @AFASbart for suggesting the cause and fix.
- Fix: #159: Swipe step is not working due to bad '??' statement
- Fix: #155: Ensure stdout reporter only add ascii colour code when the target supports it

AFASbart committed:
Fixed an issue with code generation.

Fixed issue #165

AFASbart committed:
updated isNotEmpty

jonsamwell committed:
Merge pull request #168 from AFASbart/integration_test__package_support

Fixed an issue with code generation.

jonsamwell committed:
fix(generation): fix when generating empty feature files

jonsamwell committed:
chore(libs): updated to latest dart_gherkin release

AFASbart committed:
Reversed the path change for this pull request.

AFASbart committed:
Fixed unnecessary assignment in runScenario

AFASbart committed:
Changed consistency of @deprecated

jonsamwell committed:
Merge pull request #192 from AFASbart/fixTips

Fixed unnecessary assignment in runScenario

jonsamwell committed:
Merge pull request #187 from AFASbart/fix_deprated

Code generation variables put on deprecated.

jonsamwell commented:
@joseignacioojeda I have to find the time to redo the documentation - so feel free to do a PR it would be really apprenticed.  I am struggling to find time at the moment.

Gustl22 committed:
Allow disabling handle semantics in testWidgets

jonsamwell committed:
Merge pull request #200 from golane-august/semantics-option

Allow disabling semantics in testWidgets

BartVerm779 committed:
Added path support

BartVerm779 committed:
Revert "Added path support"

This reverts commit 434db8405cec022890e847bf193f095d1c67e397.

Commit:
Fix CHANGELOG filename

Commit:
Update README.md

AFASbart committed:
Merge pull request #215 from vrnvorona/patch-1

README and Changelog have an updated classname for the flutter_gherkin.dart class

AFASbart committed:
Merge pull request #1 from jonsamwell/integration_test__package_support

update (3-5-2022)

AFASbart committed:
The test case gets retried 3 times and the other steps will get skipped

AFASbart committed:
Max attempts can be changed by the user

AFASbart committed:
Update pubspec.lock

AFASbart committed:
Update pubspec.yaml

AFASbart committed:
Merge remote-tracking branch 'origin/integration_test__package_support' into integration_test__package_support

AFASbart committed:
Lint fixes

jonsamwell committed:
Merge pull request #220 from BartVerm779/integration_test__package_support

Step retries, Scenario result and skipped steps.

AFASbart committed:
Fixed the max retries to at least have 1 try and a delay has been added

jonsamwell committed:
Merge pull request #221 from BartVerm779/integration_test__package_support

Fixed the max retries to at least have 1 try and a delay has been added

AFASbart committed:
Started on the dart_gherkin 3.0 conversion

jonsamwell committed:
feat(config): re-worked configuration so it can stay mostly immutable
- Fix #195: Adding missing export for `wait_until_key_exists_step.dart`
- Fix #226: Allow compatibility with dev and master flutter branches
- Feat #218: Allow retry steps in case of intermittent failure by setting the configuration properties `stepMaxRetries` & `retryDelay`
- Fix #210 & #191: Ability to take screenshots on web
- Fix #198: Allow the use of implicit pumpAndSettle methods in the app driver to be turned off using the configuration property `waitImplicitlyAfterAction`

jonsamwell committed:
chore(lints): fixed linting and improved code quality
fix(reporters): fixed summary reporter being called after each feature

jonsamwell committed:
fix(app): pump and settle app after test to ensure finalisation can happen

jonsamwell committed:
chore(docs): added configuration hint

jonsamwell committed:
Merge pull request #230 from jonsamwell/dart-3.0support

Dart_gherkin 3.0 conversion

jonsamwell committed:
chore(lib): remove file that should not be checked in

jonsamwell committed:
chore(version): ensure this branch is still published under a rc release

jonsamwell committed:
chore(docs): made changelog match pubspec version number

jonsamwell committed:
- Fix #231 - Removed the use of explicitly calling `pumpAndSettle` in the pre-defined steps in favour of the implicit `pumpAndSettle` calls used in the `WidgetTesterAppDriverAdapter`.
  - Added ability to add a `appLifecyclePumpHandler` to override the default handler that determines how the app is pumped during lifecycle events.  Useful if your app has a long splash screen etc. Parameter is on `executeTestSuite`.
  - Added ability to ensure feature paths are relative when generating reports `useAbsolutePaths` on the `GherkinTestSuite` attribute

* BREAKING CHANGE: The parameters on `executeTestSuite` are now keyed to allow for the above changes

jonsamwell committed:
fix(#222): escape single quotation marks in data tables

jonsamwell committed:
- Fix #235 - fix issue taking a screenshot on an Android device
  - Resolved #170: Added example code to ensure json report is save to disk even when the test run fails. Also added script to generate a HTML report from a JSON report

jonsamwell committed:
fix(#237): remove reference to dart:html and gaurd stdout function so library works on web

jonsamwell committed:
fix(#231): Exposed `frameBindingPolicy` on the test runner when running tests which can affect how frames are painted and the speed of the test run, I've removed the default value which might be responsible for #231

jonsamwell committed:
- Fix #231 - using local coordinate system when taking a screenshot on Android (thanks to @youssef-t for the solution)
  - Fix #216 - ensure step exceptions and `expect` failure results are added as errors to the json report
  - Scenarios can now have descriptions which also appear in the json reporter output

robert-virkus committed:
feat: add standard configuration for web

This eases running tests on Flutter web #243

Commit:
Fix redundant escaping

Commit:
--keep-app-running is no more avaliable

jonsamwell committed:
Merge pull request #250 from huylg/integration_test__package_support

--keep-app-running is no more avaliable

jonsamwell committed:
Merge pull request #246 from youssef-t/integration_test__package_support

- Fix #245 - removing the redundant escaping

jonsamwell committed:
Merge pull request #244 from robert-virkus/integration_test__package_support

feat: add standard configuration for web

jonsamwell committed:
- Fix #257 - fixed issue when generating a step with a '$' sign in
  - Fix #256 - Ensure all exceptions generated when running a step are logged
  - Fix #253 - Ensure features with descriptions that span more than one line are parsed correctly
  - Fix #252 - Ensure all async code is awaited
  - When taking a screenshot on the web use the render element rather than relying on native code that does not work

lsuhov committed:
Change scroll code to make it working also for AnimatedList

AnimatedList is not subclassing from ScrollView, and the seen behaviour was that nativeDriver.state was trying to ensure that a single state is available which is not true for AnimatedList. Regression test was performed on ListView.

jonsamwell committed:
Merge pull request #273 from lsuhov/integration_test__package_support

Change scroll code to make it working also for AnimatedList

AFASbart committed:
Added the possibility to use flutter test instead of flutter drive

AFASbart committed:
Updated README

AFASbart committed:
Forgot to delete 3 lines from my own project

AFASbart committed:
Update lib/src/flutter/code_generation/generators/gherkin_suite_test_generator.dart

Co-authored-by: August Oberhauser <86839043+golane-august@users.noreply.github.com>

AFASbart committed:
Update lib/src/flutter/code_generation/generators/gherkin_suite_test_generator.dart

Co-authored-by: August Oberhauser <86839043+golane-august@users.noreply.github.com>

jonsamwell committed:
Merge pull request #279 from jonsamwell/test-compatibility

Make it possible to run the tests with flutter test instead of flutter drive.

golane-august committed:
fix: remove bracket on gherkin generator

timobaehr committed:
chore(lib): analyzer <6.0.0 insteal 5.0.0

jonsamwell committed:
Merge pull request #290 from timobaehr/integration_test__package_support

Compability with Dart 3, Flutter 3.10

vrnvorona commented:
any chance to bump to rc18 on hosted?

vrnvorona committed:
Fix code generation error after b7c50bde9f63eee0a6bd3bf444ec471b063aa8e3

jonsamwell committed:
Merge pull request #292 from vrnvorona/patch-2

Fix code generation error after e8c874f0c07d7f575f74501f31f71140d7edff80

jonsamwell committed:
Merge branch 'integration_test__package_support' into fix/remove-bracket-generator

jonsamwell committed:
Merge pull request #280 from golane-august/fix/remove-bracket-generator

fix: remove bracket on gherkin generator

lsuhov committed:
Remove unnecessary_await_in_return warning from gherkin g.dart files

jonsamwell committed:
Merge pull request #305 from lsuhov/integration_test__package_support

Remove unnecessary_await_in_return warning from gherkin g.dart files

cbarajasZF commented:
Hi! @jonsamwell 
I noticed that this PR has been open for quite some time. We’ve been using this branch for more than a year now, and it has been very useful across multiple projects.

Is there any update on whether this branch will be merged?
Appreciate your work on this! This branch has been really useful for us.
CBarajas
